### ISSUE TITLE
u gate matrix representation check

### ISSUE THREAD

**Issue by zlatko-minev** [Qiskit/qiskit-terra#6435]
### What is the expected enhancement?

Perhaps some different gate definition is used, since when I check the matrix representation of the `u` gate in https://qiskit.org/documentation/stubs/qiskit.circuit.library.UGate.html, I find that the off-diagonals don't quite match up what i see in the documentation. Maybe I'm missing something simple?

![image](https://user-images.githubusercontent.com/15937254/118654724-d88d4580-b7b6-11eb-8d1d-42536b0c97b0.png)
vs.
![image](https://user-images.githubusercontent.com/15937254/118654741-db883600-b7b6-11eb-80cb-6359994bd144.png)

I obtained the above using:

![image](https://user-images.githubusercontent.com/15937254/118654770-e4790780-b7b6-11eb-8354-24083d0958f8.png)

![image](https://user-images.githubusercontent.com/15937254/118654800-ef339c80-b7b6-11eb-92c4-530c01e2cd23.png)

**Comment by singhmeet11** [comment]
I would like to work on this

**Comment by jond01** [comment]
I agree, except for the global phase (which is important), the off diagonals of
![image](https://user-images.githubusercontent.com/36337649/126503786-b17a5c46-7320-4b2e-b3d6-2521b056f126.png)
and
![image](https://user-images.githubusercontent.com/36337649/126503865-c50a48a0-e581-4f95-a4f1-90658e195d59.png)
do not concur.
I used the code in the gist below to check it:
https://gist.github.com/jond01/2a71557438d424ee41d6d5501589eada
(Specifically, check `Out[9]` and `Out[10]`.)

Should we fix the documentation? Maybe remove the first definition to avoid confusion? (the true array definition of this gate is the 2nd one)

**Comment by zlatko-minev** [comment]
I imagine the reason that the gate definition above was given in terms of Rx Rz is because that’s how it’s implemented in the backend hardware gatedef. I think I recall that there is some minus sign are as well
Not sure …

**Comment by jond01** [comment]
Ok, I think that the following is a valid explanation:

## OpenQASM 2.0
U was defined as:
https://arxiv.org/pdf/1707.03429.pdf#page=5
![image](https://user-images.githubusercontent.com/36337649/126565311-34c202a9-2ac0-43d8-9039-18def165ee4e.png)


Instead of [qiskit's wrong definition](https://qiskit.org/documentation/stubs/qiskit.circuit.library.UGate.html) (even with global phase tolerance):
![image](https://user-images.githubusercontent.com/36337649/126564915-db78d30c-a4d2-4fe3-9f71-927d97554b80.png)
The following is correct and implemented with a pair of π/2-pulses (from the paper above):
![image](https://user-images.githubusercontent.com/36337649/126565198-7398d292-761b-4f42-bd31-b65fc839c85f.png)

I verified the equality between the definitions of U in eq. (2) and (5) above (see the updated [gist](https://gist.github.com/jond01/2a71557438d424ee41d6d5501589eada) - "OpenQASM 2.0 definitions").

## OpenQASM 3.0
The definition of U gate has changed by a global phase from QASM 2.0 to 3.0:
https://arxiv.org/pdf/2104.14722.pdf#page=11
![image](https://user-images.githubusercontent.com/36337649/126526605-952539d2-1a2f-4da9-be3c-d3465ea3d44f.png)
![image](https://user-images.githubusercontent.com/36337649/126526914-250fe34b-fc95-4d73-b6cc-9f1b2eeb9337.png)

-----

In conclusion, I think that the correct equality in QASM 3.0 is:
![image](https://user-images.githubusercontent.com/36337649/126608872-876e759b-8be9-4c27-aa57-5f6fa0fa03c9.png)
And we should change it in our the documentation (below "Implemented using two X90 pulses on IBM Quantum systems").

**Comment by jond01** [comment]
See also the implmentation in `cirq`:
https://github.com/quantumlib/Cirq/blob/782c14e0cc821794d21e82f30df8e6998f4b9992/cirq-core/cirq/circuits/qasm_output.py#L75-L81
It matches the true definition:
![image](https://user-images.githubusercontent.com/36337649/126565311-34c202a9-2ac0-43d8-9039-18def165ee4e.png)

Rather than the wrong one from [`qiskit`](https://qiskit.org/documentation/stubs/qiskit.circuit.library.UGate.html):
![image](https://user-images.githubusercontent.com/36337649/126564915-db78d30c-a4d2-4fe3-9f71-927d97554b80.png)
